doc/ilenia.8.in
.TH ilenia 8 "" "ilenia #VERSION#" ""
.SH NAME
ilenia \- A package manager for CRUX (and CRUX PPC of course) 
.SH SYNOPSIS
\fBilenia ACT [OPTION(S)] [ARG(S)]\fP
.SH DESCRIPTION
ilenia is a complete and probably simple to use ports packages manager for CRUX
and CRUX PPC. With the different actions it is possible to install, update or
remove packages checking the dependencies; it is possible to search for ports by
their name or by the description field; it is possible to obtain lists or trees
of the dependencies or dependents of a port or of some ports and it is possible
to see a list of ports with different version from the ones in the ports tree or
with lower version than the ones in the ports tree.
.br
.PP
.B "The locked version, favourite repository, repositories hierarchy, higher version rule."
.br
The rule for install or update a port is the locked version, favourite
repository, repositories hierarchy, higher version one: if a port is contained
in the locked versions list it will be installed (if it is found) only with the
specified version; then if a port is contained in the favourite repositories it
will be installed or updated only from the associated repository; if the port
isn't contained by the locked versions or the favourite versions, and it is
found in more than one repository it is installed from the repository with
higher priority (that compares first in the repositories hierarchy); finally if
the port is found more than in one repository with the same priority it is
installed the one with higher version.
.SH ACTIONS
.TP
.B "\-d, \-\-diff"
Shows the list of the installed packages with different version from the ones in
the ports tree.
.TP
.B "\-D, \-\-dependencies"
Shows the list of the dependencies needed to install the port(s) in ARG(S). 
This action can be used with the "\-\-verbose" option to obtain more 
information (such as the repository where the port is found, or i the port has 
a README) for each port in the list.
To obtain the complete list of dependencies the "\-\-all" options has to be
used.
.TP
.B "\-i, \-\-info"
Shows information about the port(s) in ARG(S).
.TP
.B "\-l, \-\-list"
Shows the list of the ports from all repositories, or if some arguments are
supplied shows the list of the ports founded from the repositories in ARG(S).
.TP
.B "\-p, \-\-updated"
Shows the list of the installed packages with lower version from the ones in the
ports tree. The list is built according to the repositories hierarchy, locked
versions and favourite repository rules.
.TP 
.B "\-r, \-\-show\-readme"
Show the README of a port using the defaault pager (setted via the envirnment
variable PAGER).
.TP
.B "\-\-repository\-list"
Shows the list of the repositories used by ilenia.
.TP
.B "\-R, \-\-remove"
Removes the packages in ARG(S). Used with the "\-\-all" option removes the
packages in ARG(S) and the packages that depends from them.
.TP
.B "\-s, \-\-search"
Searches for the ARG(S) in the ports tree by name. It is allowed using wildcards
in ARG(S). It's possible using the "\-\-verbose" option to obtain more details
for founded ports.
.TP
.B "\-\-search\-desc"
Searches for the ARG(S) in the ports tree by description. It is allowed using
wildcards in ARG(S). It's possible using the "\-\-verbose" option to obtain more
details for founded ports.
.TP
.B "\-T, \-\-dependents"
Shows the list of the ports that require the port(s) in ARG(S) as
dependencies. Defaultly the list contains only installed ports (packages), it's
possible using the "\-\-all" option to obtain a complete list. This action can
be used with the "\-\-verbose" option to obtain more information (such as the
repository where the port is found, or i the port has a README) for each port in
the list.
.TP
.B "\-u, \-\-update"
Updates the repository in ARG(S) or, if no ARG is supplied, updates the entire
ports tree.
.TP
.B "\-U, \-\-update\-pkg"
Updates or installs the port(s) in ARG(S) and their dependencies. If no ARG is
supplied the entire system (all packages that have to be updated) will be
updated.
.SH OPTIONS
.TP
.B "\-\-all"
It is used with the "\-\-remove" or the "\-\-dependents" action to,
respectively, remove all dependents packages or show all dependents packages.
And with the "\-\-dependencies" action to obtain the complete list of the
dependencies.
.TP
.B "\-\-cache"
Forces the rebuild of the cache.
.TP
.B "\-\-fetch\-only"
It is used with the "\-\-update\-pkg" action, it forces to download only the
required sources of the ports. 
.TP
.B "\-\-just\-install"
It is userd with the "\-\-updatade\-pkg" action, it forces to don't update the
outdated dependencies but just to install not installed ones.
.TP
.B "\-\-no\-aliases"
Ignores the aliases list.
.TP
.B "\-\-no\-colors"
Disables the colored output.
.TP
.B "\-\-no\-favourite"
Ignores the favourite repositories list.
.TP
.B "\-\-no\-hierarchy"
Ignores the repositories hierarchy. This could be dangerous, probably.
.TP
.B "\-\-no\-locked"
Ignores the locked versions list.
.TP
.B "\-\-no\-never\-install"
Disables the never_install list.
.TP
.B "\-\-tree"
It is used with the "\-\-dependencies" or the "\-\-dependents" actions to obtain
a tree instead of a list of dependencies or dependents ports.
.TP
.B "\-\-verbose"
It is used with the "\-\-dependencies", the "\-\-dependents", the "\-\-\search"
or the "\-\-search\-desc" actions to obtain more information.
.TP
.B "\-?, \-\-help"
Shows an help screen.
.TP
.B "\-\-usage"
Shows a brief usage message.
.TP
.B "\-V, \-\-version"
Shows the version.
.SH CONFIGURATION: /etc/ilenia.conf
The configuration file is "/etc/ilenia.conf", it is composed by four sections:
"ilenia", "favourite_repositories", "locked_versions", "aliases", "pkgmk_confs".
.br
The "[ilenia]" sections contains the configuration variables, these variables
can be overrated on the fly setting environment variables. The variables are:
.TP
.B POST_PKGADD
It contains the commands to be executed after a success "pkgadd" (after the
installation of a single package). For default this variable is empty: no action
is performed after a pkgadd.
.TP
.B NOT_FOUND_POLICY
It configures the policy when a dependence port isn't found in the ports
tree. The possible values are: "Ask", the default one, that stops the execution
and ask the user if it is possible to continue; "Stop", that stops immediately
the execution; "Nevermind", that prints out a warning and ignores the not found
port and continues the execution.
.TP
.B REPOSITORIES_HIERARCHY
It configures the repositories hierarchy, the value is a list of repositories
separated by space, the first repository in the list is the repository with
higher priority.
.TP
.B NEVER_INSTALL
This field contains the list of ports that mustn't be installed also if they're
setted as dependencies of other ports.
.TP
.B ASK_FOR_UPDATE
Possible values are "Yes", the default one and "No"; setting this variable to
"Yes" before updating any package shows the list of ports to be installed or
updated and asks the user if is it possible to continue.
.TP
.B ENABLE_COLORS
It enables or disables the colored output, possible values are "Yes", the
default and "No".
.TP
.B VERBOSE
It enables the verbosity for default (like using always the "\-\-verbose"
option), possible values are "Yes" and "No", the default one.
.TP
.B ENABLE_XTERM_TITLE
It enables the setting informations in the xterm title bar, possible values are
"Yes" and "No", the default is "Yes".
.TP
.B DEFAULT_XTERM_TITLE
It is the default xterm title, the one to restore (if ENABLE_XTERM_TITLE is
setted to "Yes")  when ilenia exits. The default is $LOGNAME@$HOSTNAME:$PWD.
.TP
.B ENABLE_LOG
It enables the simple log for tracking install/update/remove actions; the
logfile is /var/log/ilenia. Possible values are "Yes" and "No", the default is
"Yes".
.TP
.B REJECTED_POLICY 
It configures the policy in managing rejected files while updating
packages. Possible values are "Check", to check how many rejected files there
are, "Run" to run rejmerge(8) if there are some rejected files, "Nevermind" to
disable the check for rejected files.
.PP
The section "[favourite_repositories]" configures the favourite repositories; a
port in this list can be updated or installed only from the associated
repository. The syntax is a couple of port name, repository name 
("port = repository") separated by newline, it is allowed using wildcards in
port name. For example:
.br
.br
[favourite_repositories]
.br
gcc = core
.br
xorg-* = xorg
.br
.PP
The section "[locked_versions]" configures the ports to be installed or updated
only at associated version. The syntax is a couple of port name, version
("port = version") separated by newline. For example:
.br
.br
[locked_versions]
.br
ilenia = 3.0-1
.br
glibc = 2.3.6-2
.br
.PP
The section "[aliases]" configures the aliases list; a port that is an alias of
another one has the same functions and can be substituted to this in
dependencies, for example it is possible to use "xorg" as an alias for "x11":
the ports that requires "x11" works (almost always) also if it is installed
"xorg". The syntax is like the ones of the "favourite_repositories" or the
"locked_versions" sections:
.br
.br
[aliases]
.br
x11 = xorg
.br
sendmail = exim postfix
.br
.PP
The section "[pkgmk_confs]" configures the ports to be built using the
non-default pkgmk configuration file (/etc/pkgmk.conf). The syntax is: 
"port = /path/to/pkgmk.conf"; it is allowed using wildcard in port's names.
.SH LOCAL REPOSITORIES
.PP
Local repositories are repositories that resides only on the local machine
created by the local sysadmin. To let ilenia recognize the admin have to create
a file in "/etc/ports/" with ".local" extension (for example: my.local); this
file must be a single row, where there is specified the full path of the local
repository. 
It's possible to update automatically (i.e.: downloading them from various 
repositories) adding a shell script called ".updater.sh" in the same directory
where ports reside; this shell script will be executed in the repository's
directory. For example:
.br
# my.local
.br
.br
PATH="/root/ports"
.br
.br
# End of file
.br
.br
#!/bin/bash
.br
#
.br
# /root/ports/.updater.sh
.br
#
.br
rsync -aqz cruxppc.org::ports/core/2.5/ilenia/ ilenia
.br
# End Of File
.br
.SH SEE ALSO
pkgadd(8), pkgrm(8), pkginfo(8), rsync(1), httpup(8)
.SH AUTHOR
.nf
Coviello Giuseppe <cjg@cruxppc.org>
http://ilenia.berlios.de
.fi


NEWS
25/02/2007 - ilenia-3.4 released 
Programming ilenia  is like researching the  philosophal stone: I've  used a new
data  structure  in this  release  that made  ilenia  run  faster while  solving
dependencies, it's about the 4% faster.  
This is also a bugfix release  (like all others, of course): it's solved another
of the nullpointer's bug: using the deptree option to obtain the dependencies of
a port caused sometimes a segmentation  fault; and finally the count of rejected
files is correct.

22/09/2007 - ilenia-3.3 released 
Here  we  are  at  Pianeta  Amiga  in Empoli  (Florence),  and  we're  releasing
ilenia-3.3.  
The new  features, the ones visible by  the users, are the  never intall system,
that  make possibible  to config  ilenia to  never install  a list  of specified
ports, and the other new feature is the cyclic dependecies check.  
ilenia-3.3 comes  out with some  bugfixes, like the  one that made  ilenia fault
when managing ports that had the  same value in the name and description fields.
ilenia-3.3 has a  new simpler system for manage dependencies  of lists of ports,
that make ilenia runs faster.

18/08/2007 - ilenia-3.2 released 
Someone goes to the beach some goes to the mountain, everyone goes on holiday; I
make a new release!  
ilenia-3.2 comes out  with some interesting features (and  some less interesting
ones,    of   course);    ilenia-3.2    has   a    log    system   to    monitor
installing/upgrading/removing of  ports, like the other features  the log system
can be enable or disabled via the configuration file.  
Another  new  feature  of ilenia-3.2  is  the  check  for rejected  files  while
installing packages;  after an  installation or an  upgrade ilenia can  look for
rejected files and warn the user and optionally run rejmerge.  
A long time  ago Biscuttiello said that he  had to install a port  but he didn't
want to  update the  dozens dependencies  of that port,  and he  asked me  for a
feature that  can disable  the updating of  dependencies while installing  a new
package; I  thought that a feature  like that can be  potentially dangerous, and
also the port to install wouldn't compile or work if it depends on new features,
but Biscuttiello is the master of his system (like any user is the mastes of his
system) and ilenia is his servant, and he  says what he do, and then i added the
'--just-install'  option  that makes  ilenia  to just  install  a  port and  his
dependencies and doesn't update outdated ones.  
Finally the  last new  feature of ilenia-3.2  is the  setting on the  xterm (and
compatibles) title  bar a  message displaying the  action that ilenia  is doing;
this feature can be disabled or enabled via the configuration file.  
Happy holiday!

05/05/2007 - ilenia-3.1 released 
After  a  month  and  a  day  a  new release.   ilenia-3.1  has  a  new  feature
(--show-readme)  to  show,  using  your  favourite  pager,  the  README  of  the
port. There  are also some  little bugfixes: it  is fixed the  segmentation faul
caused during a system update when a dependece port isn't found; it is fixed the
segmentation fault when trying to obtain information about an "orphaned" port (a
port that is installed but it isn't found in any repository).

04/04/2007 - ilenia-3.0 released 
Fine, I'll stop to backport new features to ilenia-2.*: ilenia-3.0 is ready!  
ilenia-3.0  comes with  many new  features like  the colored  output;  the trees
output  for show the  dependencies; the  search by  description command  and the
support of  wildcards in the search  command; the support for  shell commands in
version fields of Pkgfile.  
ilenia-3.0  uses a new  unique configuration  file: /etc/ilenia.conf;  right the
user   has   to   port   the   configurations   contained   by   /etc/ilenia.rc,
/etc/ports/{aliases,favoriterepo,favoriteversion,favoritepkgmk}.  
ilenia-3.0 runs faster than ilenia-2.* and ilenia-1.* and it is more precise and
secure than the precedent versions.

27/01/2007 - ilenia-2.2 released 
Probably this will be  really the last release of ilenia 2.*:  ilenia 3 works at
50%.  After about three  years of  repositories anarchy,  where only  the higher
version was the priority to use a port from a repository instead of another one,
ilenia 2.2  introudes the repositories  hierachy available trhough  the variable
REPOS_HIERACHY in  ilenia.rc and the option "--no-repos-hierachy"  to disable at
run-time the  hierachy.  Another new  feature of ilenia  2.2 is the  checking of
outdated dependencies while installing or updating a package.

17/01/2007 - ilenia-2.1 released 
Probably this will  be the last release  of ilenia 2.*: I'm woking  on ilenia 3.
In  this  release the  are  two  new features:  the  "--info"  action to  obtain
informations about  a port  (or some ports)  and the "--fetch-only"  option used
with the "-U"  ("--udpdate-pkg") action forces ilenia to  just fetch the sources
(something like 'pkgmk -do').

21/12/2006 - ilenia-2.0.1 released
This is essentially a bug fix release: solved bugs about:
- Segmentation fault while managing favoriteversion
- Executing pre-install and post-install script not marked as executable files 
- Managing supfiles with more than a dot in their names (for example ppc.*.cvs)
- Working with alias

11/03/2006 - ilenia-2.0 released 
There are  many new things in  this release: beginning from  the sourcecode that
had been rewritten  to have a better  usage of memory and an  "unified error and
warning  message system",  all error  and warning  messages are  managed  by two
function.  The more tangible, for the user, new features are: 
- the "not found policy", when a  dependence isn't found ilenia can act in three
  ways  asking you  if she  can  procede (setting  this variable  as "Ask",  the
  default), stopping  the build  and install process  (setting this  variable as
  "Stop") or contnuing with next package (setting the value "Nevermind"); 

src/ilenia.c
/* ilenia.c */

/* ilenia -- A package manager for CRUX
 *
 * Copyright (C) 2006 - 2007
 *     Giuseppe Coviello <cjg@cruxppc.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

#include <stdio.h>
#include <argp.h>
#include <stdlib.h>
#include <string.h>
#include "../config.h"
#include "port.h"
#include "package.h"
#include "dependencies.h"
#include "conf.h"
#include "cache.h"
#include "driver.h"
#include "output.h"
#include "info.h"
#include "update.h"
#include "str.h"
#include "memory.h"
#include "remove.h"
#include "hash.h"

const char *argp_program_version = "ilenia " VERSION " (Philosophal Stone)"
    "\n"
    "Copyright (C) 2006-2007 Giuseppe Coviello.\n"
    "This is free software.  You may redistribute copies of it under the terms of\n"
    "the GNU General Public License <http://www.gnu.org/licenses/gpl.html>.\n"
    "There is NO WARRANTY, to the extent permitted by law.";
const char *argp_program_bug_address = "Giuseppe Coviello <cjg@cruxppc.org>";
static char doc[] = "A package manager for CRUX (and CRUX PPC of course)";
static char args_doc[] = "ACT [ARG(S)]";

enum OPT { OPT_CACHE = 300, OPT_REPOSITORY_LIST, OPT_NO_FAVOURITE_REPOSITORIES,
	   OPT_NO_LOCKED_VERSIONS, OPT_ALL, OPT_FETCH_ONLY, OPT_JUST_INSTALL,
	   OPT_NO_REPOSITORIES_HIERARCHY, OPT_TREE, OPT_NO_ALIASES,
	   OPT_NO_COLORS, OPT_NO_NEVER_INSTALL
};

enum ACT { ACT_UPDATE = 300, ACT_LIST, ACT_SEARCH, ACT_INFO, ACT_README,
	ACT_DIFF, ACT_UPDATED, ACT_DEPENDENCIES, ACT_UPDATE_PKG,
	ACT_DEPENDENTS, ACT_REMOVE, ACT_REPOSITORY_LIST, ACT_SEARCHDESC
};

static struct argp_option options[] = {
	{"ACTIONS", 0, 0, OPTION_DOC, "", 0},
	{"update", 'u', 0, 0, "Update ports tree", 1},
	{"list", 'l', 0, 0, "List ports", 1},
	{"search", 's', 0, 0, "Search for ports by name", 1},
	{"search-desc", ACT_SEARCHDESC, 0, 0,
	 "Search for ports by description", 1},
	{"info", 'i', 0, 0, "Get info on a port", 1},
	{"show-readme", 'r', 0, 0, "Show the README of a port", 1},
	{"diff", 'd', 0, 0, "List version differences", 1},
	{"updated", 'p', 0, 0, "List ports with newer version", 1},
	{"depedencies", 'D', 0, 0, "List dependencies of a package", 1},
	{"update-pkg", 'U', 0, 0,
	 "Update or install a package with dependencies", 1},
	{"dependents", 'T', 0, 0, "List dependents of a package", 1},
	{"remove", 'R', 0, 0, "Remove a package checking dependencies", 1},
	{"\n  OPTIONS", 0, 0, OPTION_DOC, "", 2},
	{"tree", OPT_TREE, 0, 0, "Dump a tree dependencies insted of a list",
	 3},
	{"cache", OPT_CACHE, 0, 0, "Rebuild the cache", 3},
	{"repository-list", OPT_REPOSITORY_LIST, 0, 0,
	 "List repository that ilenia are using", 1},
	{"no-favourite", OPT_NO_FAVOURITE_REPOSITORIES, 0, 0,
	 "Ignore the user's favourite repositories", 3},
	{"no-locked", OPT_NO_LOCKED_VERSIONS, 0, 0,
	 "Ignore the user's favorite versions", 3},
	{"no-colors", OPT_NO_COLORS, 0, 0, "Disable colored output", 3},
	{"all", OPT_ALL, 0, 0, "Shows or remove all dependents packages", 3},
	{"fetch-only", OPT_FETCH_ONLY, 0, 0, "Just fetch the needed sources",
	 3},
	{"just-install", OPT_JUST_INSTALL, 0, 0, 
	 "Just install dependencies don't update outdate ones"},
	{"no-hierarchy", OPT_NO_REPOSITORIES_HIERARCHY, 0, 0,
	 "Do not use the repositories gerarchy", 3},
	{"no-aliases", OPT_NO_ALIASES, 0, 0, "Do no use aliases", 3},
	{"verbose", 'v', 0, 0, "Generate verbose output", 3},
	{"no-never-install", OPT_NO_NEVER_INSTALL, 0, 0, 
	 "Avoid the never_install list", 3},
	{0}
};

struct arguments {
	int action;
	int no_favourite_repositories;
	int no_locked_versions;
	int no_aliases;
	int no_colors;
	int no_deps;
	int no_never_install;
	int all;
	int fetch_only;
	int just_install;
	int no_repositories_hierarchy;
	int tree;
	int verbose;
	int rebuild_cache;
	list_t *args;
};

static error_t parse_opt(int key, char *arg, struct argp_state *state);

static struct argp argp = { options, parse_opt, args_doc, doc };

int main(int argc, char **argv)
{
	unsigned i;
	conf_t *conf;
	dict_t *repositories;
	hash_t *ports_hash;
	dict_t *not_founds;
	list_t *ports_list;
	list_t *packages;
	list_t *list;
	list_t *drivers;
	struct arguments arguments;

	arguments.action = 0;
	arguments.no_favourite_repositories = 0;
	arguments.no_locked_versions = 0;
	arguments.no_aliases = 0;
	arguments.no_colors = 0;
	arguments.no_deps = 1;
	arguments.no_never_install = 0;
	arguments.all = 0;
	arguments.fetch_only = 0;
	arguments.just_install = 0;
	arguments.no_repositories_hierarchy = 0;
	arguments.tree = 0;
	arguments.verbose = 0;
	arguments.rebuild_cache = 0;
	arguments.args = list_new();

	if (argc < 2) {
		char *fake_arg[2];
		fake_arg[0] = "ilenia";
		fake_arg[1] = "--help";
		argp_parse(&argp, 2, fake_arg, 0, 0, &arguments);
	}

	argp_parse(&argp, argc, argv, 0, 0, &arguments);
	debug("arguments parsed!");

	conf = conf_init();
	debug("configuration initialized!");

	if (arguments.no_repositories_hierarchy) {
		list_free(conf->repositories_hierarchy, free);
		conf->repositories_hierarchy = list_new();
	}

	if (arguments.no_favourite_repositories) {
		dict_free(conf->favourite_repositories, NULL);
		conf->favourite_repositories = dict_new();
	}

	if (arguments.no_locked_versions) {
		dict_free(conf->locked_versions, free);
		conf->locked_versions = dict_new();
	}

	if (arguments.no_aliases) {
		aliases_free(conf->aliases);
		conf->aliases = dict_new();
	}

	if (arguments.no_never_install) {
		list_free(conf->never_install, free);
		conf->never_install = list_new();
	}

	drivers = drivers_list_init();
	repositories = repositories_dict_init(drivers,
					      conf->repositories_hierarchy);
	if (!conf->enable_colors || arguments.no_colors)
		cprintf = uncoloredprintf;
	else
		cprintf = coloredprintf;

	if (arguments.rebuild_cache)
		cache_build(repositories, conf->enable_xterm_title);
	packages = packages_list_init();
	if ((ports_list = ports_list_init(repositories,
					  conf->enable_xterm_title)) == NULL) {
		list_free(arguments.args, NULL);
		dict_free(repositories, repository_free);
		list_free(packages, port_free);
		list_free(drivers, free);
		dict_free(conf->favourite_repositories, free);
		conf->favourite_repositories = dict_new();
		conf_free(conf);
		return 1;
	}

	conf_reparse(conf, ports_list);
#ifndef NDEBUG
	conf_dump(conf);
#endif
	ports_hash = port_hash_init(ports_list, packages, conf);
	not_founds = dict_new();

	if (!arguments.verbose)
		arguments.verbose = conf->verbose;

	switch (arguments.action) {
	case ACT_LIST:
		if (!arguments.args->length) {
			list_dump(ports_list, port_dump);
			break;
		}
		for (i = 0; i < arguments.args->length; i++) {
			if (dict_get(repositories, list_get(arguments.args, i))
			    == NULL) {
				warning("repository %s not found!",
					list_get(arguments.args, i));
				continue;
			}
			list = list_query(ports_list,
					  port_query_by_repository,
					  list_get(arguments.args, i));
			list_dump(list, port_dump);
			list_free(list, NULL);
		}
		break;
	case ACT_SEARCH:
		if (!arguments.args->length) {
			warning("action --search (-s) requires at least an "
				"argument!");
			break;
		}
		for (i = 0; i < arguments.args->length; i++) {
			list = list_query(ports_list, port_query_by_name,
					  list_get(arguments.args, i));
			if (arguments.verbose)
				list_dump(list, port_info_dump);
			else
				list_dump(list, port_dump);
			if (list->length == 0)
				printf("%s not found!\n",
				       (char *)list_get(arguments.args, i));
			list_free(list, NULL);
		}
		break;
	case ACT_SEARCHDESC:
		if (!arguments.args->length) {
			warning("action --search-desc requires at least an "
				"argument!");
			break;
		}
		for (i = 0; i < arguments.args->length; i++) {
			char *key = xstrdup(list_get(arguments.args, i));
			if (isalpha(*key) && isalpha(*(key + strlen(key) - 1))) {
				strprepend(&key, "*");
				strappend(&key, "*");
			}
			list = list_query(ports_list, port_query_by_description,
					  key);
			if (arguments.verbose)
				list_dump(list, port_info_dump);
			else
				list_dump(list, port_dump);
			list_free(list, NULL);
			free(key);
		}
		break;
	case ACT_DIFF:
		port_show_diffs(ports_hash, packages, conf->enable_xterm_title);
		break;
	case ACT_UPDATED:
		port_show_outdated(ports_hash, packages, conf->enable_xterm_title);
		break;
	case ACT_DEPENDENCIES:
		if (!arguments.args->length) {
			warning("action --dependencies (-D) requires at "
				"least an argument!");
			break;
		}
		dependencies_dump(arguments.args,
				  ports_hash, conf->aliases,
				  not_founds,
				  arguments.tree, arguments.verbose, 
				  conf->enable_xterm_title, arguments.all);
		break;
	case ACT_DEPENDENTS:
		if (!arguments.args->length) {
			warning("action --dependents (-T) requires at "
				"least an argument!");
			break;
		}
		for (i = 0; i < arguments.args->length; i++)
			dependents_dump(list_get(arguments.args, i),
					ports_hash, conf->aliases,
					arguments.tree, arguments.verbose,
					arguments.all, conf->enable_xterm_title);
		break;
	case ACT_INFO:
		if (!arguments.args->length) {
			warning("action --info requires at "
				"least an argument!");
			break;
		}
		for (i = 0; i < arguments.args->length; i++)
			info_dump(list_get(arguments.args, i), ports_hash);
		break;
	case ACT_README:
		if (!arguments.args->length) {
			warning("action --show-readme requires at "
				"least an argument!");
			break;
		}
		for (i = 0; i < arguments.args->length; i++)
			readme_dump(list_get(arguments.args, i), ports_hash);
		break;
	case ACT_REPOSITORY_LIST:
		repositories_dict_dump(repositories);
		break;
	case ACT_UPDATE:
		if (arguments.args->length)
			repositories_dict_update(repositories, arguments.args,
						 conf->enable_xterm_title); 
		else
			repositories_dict_update_all(repositories,
				conf->enable_xterm_title);
		break;
	case ACT_REMOVE:
		if (!arguments.args->length) {
			warning("action --remove (-R) requires at "
				"least an argument!");
			break;
		}
		remove_packages(arguments.args, packages, ports_hash, conf,
				arguments.all);
		break;
	case ACT_UPDATE_PKG:
		if (!arguments.args->length) {
			update_system(ports_hash,
				      arguments.fetch_only,
				      conf);
			break;
		}
		update_package(arguments.args, ports_hash,
			       arguments.fetch_only,
			       conf, arguments.just_install);
		break;
	default:
		if (!arguments.rebuild_cache)
			warning("What can I do for you?");
	}

	if (conf->enable_xterm_title)
		xterm_reset_title(conf->default_xterm_title);

	list_free(arguments.args, NULL);
	dict_free(repositories, repository_free);
	list_free(ports_list, port_free);
	list_free(packages, port_free);
	hash_free(ports_hash, NULL);
	list_free(drivers, free);
	conf_free(conf);
	dict_free(not_founds, port_free);

	return 0;
}

static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
	struct arguments *arguments = state->input;

	switch (key) {
	case 'u':
		arguments->action = ACT_UPDATE;
		break;
	case 'l':
		arguments->action = ACT_LIST;
		break;
	case 's':
		arguments->action = ACT_SEARCH;
		break;
	case 'i':
		arguments->action = ACT_INFO;
		break;
	case 'r':
		arguments->action = ACT_README;
		break;
	case 'd':
		arguments->action = ACT_DIFF;
		break;
	case 'p':
		arguments->action = ACT_UPDATED;
		break;
	case 'D':
		arguments->action = ACT_DEPENDENCIES;
		break;
	case 'U':
		arguments->action = ACT_UPDATE_PKG;
		break;
	case 'T':
		arguments->action = ACT_DEPENDENTS;
		break;
	case 'R':
		arguments->action = ACT_REMOVE;
		break;
	case ACT_SEARCHDESC:
		arguments->action = ACT_SEARCHDESC;
		break;
	case OPT_CACHE:
		arguments->rebuild_cache = 1;
		break;
	case OPT_REPOSITORY_LIST:
		arguments->action = ACT_REPOSITORY_LIST;
		break;
	case OPT_NO_FAVOURITE_REPOSITORIES:
		arguments->no_favourite_repositories = 1;
		break;
	case OPT_NO_LOCKED_VERSIONS:
		arguments->no_locked_versions = 1;
		break;
	case OPT_ALL:
		arguments->all = 1;
		break;
	case OPT_FETCH_ONLY:
		arguments->fetch_only = 1;
		break;
	case OPT_JUST_INSTALL:
		arguments->just_install = 1;
		break;
	case OPT_NO_REPOSITORIES_HIERARCHY:
		arguments->no_repositories_hierarchy = 1;
		break;
	case OPT_TREE:
		arguments->tree = 1;
		break;
	case OPT_NO_ALIASES:
		arguments->no_aliases = 1;
		break;
	case OPT_NO_COLORS:
		arguments->no_colors = 1;
		break;
	case OPT_NO_NEVER_INSTALL:
		arguments->no_never_install = 1;
		break;
	case 'v':
		arguments->verbose = 1;
		break;
	case ARGP_KEY_ARG:
		list_append(arguments->args, arg);
		break;
	default:
		return ARGP_ERR_UNKNOWN;
	}

	return 0;
}
